tidy3d/components/data/sim_data.py

```diff
@@ -971,34 +979,244 @@ def make_adjoint_sim(
             monitors=adjoint_monitors,
         )
 
+        if not norm_source:
+            sim_adj_update_dict["normalize_index"] = None
+
         # set the ADJ grid spec wavelength to the original wavelength (for same meshing)
         grid_spec_original = sim_original.grid_spec
         if sim_original.sources and grid_spec_original.wavelength is None:
             wavelength_original = grid_spec_original.wavelength_from_sources(sim_original.sources)
             grid_spec_adj = grid_spec_original.updated_copy(wavelength=wavelength_original)
             sim_adj_update_dict["grid_spec"] = grid_spec_adj
 
-        return sim_original.updated_copy(**sim_adj_update_dict)
+        return sim_original.updated_copy(**sim_adj_update_dict), post_norm
 
-    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> list[Source]:
+    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> dict[str, Source]:
         """Generate all of the non-zero sources for the adjoint simulation given the VJP data."""
 
-        # TODO: determine if we can do multi-frequency sources
-
         # map of index into 'self.data' to the list of datasets we need adjoint sources for
         adj_src_map = defaultdict(list)
         for _, index, dataset_name in data_vjp_paths:
             adj_src_map[index].append(dataset_name)
 
-        # gather a list of adjoint sources for every monitor data in the VJP that needs one
-        sources_adj_all = []
+        # gather a dict of adjoint sources for every monitor data in the VJP that needs one
+        sources_adj_all = defaultdict(list)
         for data_index, dataset_names in adj_src_map.items():
             mnt_data = self.data[data_index]
-            sources_adj = mnt_data.make_adjoint_sources(dataset_names=dataset_names)
-            sources_adj_all += sources_adj
+            sources_adj = mnt_data.make_adjoint_sources(
+                dataset_names=dataset_names, fwidth=self.fwidth_adj
+            )
+            sources_adj_all[mnt_data.monitor.name] = sources_adj
 
         return sources_adj_all
 
+    @staticmethod
+    def get_amp(src_time: GaussianPulse) -> complex:
+        """grab the complex amplitude from a ``SourceTime``."""
+        mag = src_time.amplitude
+        phase = np.exp(1j * src_time.phase)
+        return mag * phase
+
+    @staticmethod
+    def set_amp(src_time: GaussianPulse, amp: complex) -> GaussianPulse:
```

Comment:
Bit nitpicky but this is not really `set`ting anything. It's more like a `from_amp`?

Reply:
yea, I even considered putting these methods inside of `GuassianPulse` directly.. maybe I should

Reply:
Added `GaussianPulse.amp_complex` property as well as `GaussianPulse.from_amp_complex()` class method constructor and modified the code to use these.